### PR TITLE
Refactor handlers for rest/info and server metadata

### PR DESCRIPTION
**Pull request by rgwozdz**
This is a refactor for the code that generates the response body for the `/rest/info` and `/FeatureServer` endpoints.  Much of is just moving code and reorganizing file structure into a more coherent pattern.  Unit tests added.

### CODE REVIEW

File: lib/helpers/normalize-extent.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  const { value: cornerArrayExtent } = validate(input, cornerArraySchema)
+
+  if (cornerArrayExtent) {
+    return cornerArrayToEsriExtent(cornerArrayExtent, spatialReference)
+  }
+
+  const { value: esriExtent } = validate(input, esriExtentSchema)
+
+  if (esriExtent) {
+    return { ...esriExtent, spatialReference }
```
Comment (haoliangyu) [inline lib/helpers/normalize-extent.js:36]: If `esriExtent.spatialReference` exists and `spatialReference` is undefined, the new extent will have undefined spatial reference. Does it matter?

File: lib/server-info-route-handler.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  if (type === 'FeatureCollection') {
+    const geometryType = getGeometryTypeFromGeojson(input)
+    if (geometryType) return { layers: [input], tables }
+    return { tables: [input], layers }
+  }
+
+  // To ensure tables and layers have been properly segregated merge and re-bin
+  const merged = _.concat(layers, tables)
+
+  return categorizeTableAndLayers(merged)
```
Comment (haoliangyu) [inline lib/server-info-route-handler.js:45]: Why do you combine the array and re-do the categorization?

File: lib/server-info-route-handler.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+function normalizeInput (input) {
+  const { type, tables = [], layers = [] } = input
+
+  if (type === 'FeatureCollection') {
+    const geometryType = getGeometryTypeFromGeojson(input)
+    if (geometryType) return { layers: [input], tables }
+    return { tables: [input], layers }
+  }
```
Comment (haoliangyu) [inline lib/server-info-route-handler.js:40]: Why is the original `tables` or `layers` array overwritten?

File: lib/server-info-route-handler.js (hunk 2 above)
Comment (rgwozdz) [reply lib/server-info-route-handler.js:45]: Thanks for asking. There is no guarantee that the incoming `layers` and `tables` have been properly segregated (anything without a geometryType should be a table).  This acts as a correction for uniformed users that pass all collections in `layers` even if the data has no geometry.  In fact, we had an integration test that passed in a tabular collection in the `layers` array, with the resulting response being wrong (empty `tables: []`).

I'm chose to re-do categorization in an effort to simplify `getData` operations, but we could instead just do a validation an either reject or warn when users wrongly put tabular data into the `layers` input array.

File: lib/server-info-route-handler.js (hunk 3 above)
Comment (rgwozdz) [reply lib/server-info-route-handler.js:40]: see previous comment.

File: lib/helpers/normalize-extent.js (hunk 1 above)
Comment (rgwozdz) [reply lib/helpers/normalize-extent.js:36]: I think it does matter.  I will look.

Comment (rgwozdz) [reply lib/helpers/normalize-extent.js:36]: Same pattern exists in the pre-refactored code.  However, I would call it a bug.  Will fix.

**Review by haoliangyu — APPROVED** [review]
Thanks for the answer. We can put the bug fix in the next PR and call this done.

**Pull request merged**